src/gradio_ui.py

```diff
@@ -100,6 +100,46 @@ def _regenerate(
         def _undo_last_exchange(model_history, gradio_history):
             '''Undoes the last exchange (message pair).'''
             return model_history[:-2], gradio_history[:-1], gradio_history[:-1]
+            
+        def _save_chat_history(gradio_history, *char_setting_states):
+            '''Saves the current chat history to a .json file.'''
+            char_name = char_setting_states[0]
+            with open(f"{char_name}_conversation.json", "w") as f:
+                f.write(json.dumps({"chat": gradio_history}))
+            return f"{char_name}_conversation.json"
+            
+                
+        def _load_chat_history(file_obj, *char_setting_states):
+            '''Loads up a chat history from a .json file.'''
+            file_data = json.loads(file_obj.decode('utf-8'))
+            gradio_history = file_data["chat"]
+            char_name = char_setting_states[0]
+            
+            # If the chat is more than 4 turns we grab only the last 4 for the model history
+            model_history = gradio_history[-4:] if len(gradio_history) >= 4 else gradio_history
+            # Format the gradio history into a new model history
+            new_model_history = []
+            for line in itertools.chain.from_iterable(model_history):
+                # Chop off "\n" from the end of the line
+                formatted_line = line[:-1]
```

Comment:
Stripping off `\n` is usually done with `.strip()` or `.rstrip()`

Reply:
Changed, will add in future commit